### ISSUE TITLE
Keyword 'DICM' not found in the header (file offset 132)

### ISSUE THREAD

**Issue by wkoszek** [suyashkumar/dicom#59]
# Problem

I have an XRay DICOM failing to initialize the parser b/c of:

```
Keyword 'DICM' not found in the header (file offset 132)
```

I can dump it with with fine `dcmtk` [dcmdump]. Good news is that the competition fails as well:

```
Traceback (most recent call last):
  File "../bench/dicom_bench.py", line 29, in <module>
    ds = pydicom.dcmread(fn)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pydicom/filereader.py", line 870, in dcmread
    force=force, specific_tags=specific_tags)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pydicom/filereader.py", line 667, in read_partial
    preamble = read_preamble(fileobj, force)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pydicom/filereader.py", line 620, in read_preamble
    raise InvalidDicomError("File is missing DICOM File Meta Information "
pydicom.errors.InvalidDicomError: File is missing DICOM File Meta Information header or the 'DICM' prefix is missing from the header. Use force=True to force reading.
```

**Comment by suyashkumar** [comment]
Hm, this is an interesting one. Checking for this magic word is a good way to ensure you're actually reading a DICOM and that it isn't malformed. One option we can provide is a `force` option or flag that will not hard fail on this check, but might fail in other ways downstream since we don't have this magic word.

**Comment by IChocked** [comment]
@suyashkumar I've provided some test files from the [dicom4android](https://github.com/dragonlet/dicom4android) repo that I was able to reproduce this error with. They appear to have problems in addition to this error (for example, some are missing the header data and thus would invoke [issue 53](https://github.com/suyashkumar/dicom/issues/53) because that tag wouldn't be there). However, we have DICOMs that conform to the standard except in the way described in this issue.

[issue59.zip](https://github.com/suyashkumar/dicom/files/4833973/issue59.zip)

**Comment by IChocked** [comment]
@suyashkumar As a potential fix like you described above, what do you think about:
1. Adding a force boolean to the [parser struct](https://github.com/suyashkumar/dicom/blob/master/parse.go#L34) which can be set through a setter function (or directly, whichever is preferred)
2. Adding a default value of the force boolean as false with something like [this implementation](https://www.geeksforgeeks.org/how-to-assign-default-value-for-struct-field-in-golang/)
3. Adding an `&& p.force == false` case the [if statement](https://github.com/suyashkumar/dicom/blob/master/parse.go#L469) that throws the 'header not found' error, or some logic to log it as a non fatal error and not return. For example, inside of the if statement above, something like:
`
if s := p.decoder.ReadString(4); s != "DICM" {
       if p.force {
              log error and continue
        } else {
             p.decoder.SetError(errors.New("Keyword 'DICM' not found in the header"))
	     return nil
       }
}`

**Comment by suyashkumar** [comment]
@IChocked this approach seems reasonable to me. 

Regarding the default false value, the [zero values for booleans is false anyway](https://tour.golang.org/basics/12#:~:text=The%20zero%20value%20is%3A,the%20empty%20string)%20for%20strings.), so just adding it to the struct should be fine.  

Note the following code after that if statement expects that the other preamble tags appear in a certain order. If this is not the case, that may lead to failures as well.

@IChocked are you interested in submitting a PR for this? If not I can take care of it as well!

**Comment by IChocked** [comment]
@suyashkumar Yeah, I'd definitely be interested in submitting a PR for this! I'll get on it and submit something as soon as I can

**Comment by suyashkumar** [comment]
@IChocked do you have a data citation for the files you uploaded earlier? I'd like to include them in the testfiles for this repo if possible :).

I think I have a semi-elegant fix for this.

**Comment by suyashkumar** [comment]
cc: @wkoszek in case you may have the citation handy

**Comment by wkoszek** [comment]
@suyashkumar We don't know where https://github.com/dragonlet/dicom4android took them from. I found this repo by accident, and it happened to have the exact same problem as the problem we had experienced in production.

**Comment by suyashkumar** [comment]
Closing this out due to the fixes in #127. Let me know if we're missing anything here!
